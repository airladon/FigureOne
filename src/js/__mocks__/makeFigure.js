// import {
//   FigureElementPrimitive,
//   FigureElementCollection,
//   // AnimationPhase,
// } from '../Element';
import Figure from '../figure/Figure';
import {
  Rect, getPoint, Point, getRect,
} from '../tools/g2';
import webgl from './WebGLInstanceMock';
import DrawContext2D from './DrawContext2DMock';
import * as tools from '../tools/tools';
import { round } from '../tools/math';

tools.isTouchDevice = jest.fn();

jest.mock('../figure/Gesture');
jest.mock('../figure/webgl/webgl');
jest.mock('../figure/DrawContext2D');
jest.mock('../figure/DrawingObjects/HTMLObject/HTMLObject');

const generateRandomStringMock = jest.spyOn(tools, 'generateRandomString');
generateRandomStringMock.mockImplementation(() => '000000');

export default function makeFigure(
  inputCanvasIn = new Rect(100, -300, 1000, 500),
  scene = {},
) {
  document.body.innerHTML =
    '<div id="c">'
    + '  <canvas class="figureone__gl" id="id_figureone__gl__low">'
    + '  </canvas>'
    + '  <canvas class="figureone__text" id="id_figureone__text__low">'
    + '  </canvas>'
    + '  <div class="figureone__html">'
    + '  </div>'
    + '  <canvas class="figureone__gl" id="id_figureone__gl__high">'
    + '  </canvas>'
    + '  <canvas class="figureone__text" id="id_figureone__text__high">'
    + '  </canvas>'
    + '</div>';
  // canvas = document.getElementById('c');
  const inputCanvas = getRect(inputCanvasIn);
  // const inputLimits = getRect(inputLimitsIn);
  const definition = {
    width: inputCanvas.width,
    height: inputCanvas.height,
    limits: new Rect(-1, -1, 2, 2),
  };

  const canvasMock = {
    width: definition.width,
    height: definition.height,
    // offsetLeft: 100,
    left: inputCanvas.left,
    // offsetTop: 200,
    top: inputCanvas.top,
    // width: definition.width,
    // height: definition.height,
    offsetWidth: definition.width,
    offsetHeight: definition.height,
    scrollLeft: 0,
    scrollTop: 0,
    // eslint-disable-next-line arrow-body-style
    getBoundingClientRect: () => {
      return {
        left: inputCanvas.left,
        top: inputCanvas.top,
        width: definition.width,
        height: definition.height,
      };
    },
  };
  const htmlCanvasMock = {
    style: {
      fontsize: 1,
    },
    offsetWidth: 100,
    appendChild: () => {},
  };
  const { limits } = definition;
  const figure = new Figure({
    htmlId: 'c', limits, color: [1, 0, 0, 1], scene,
  });
  figure.webglLow = webgl;
  figure.webglHigh = webgl;
  figure.webgl = webgl;
  figure.canvasLow = canvasMock;
  figure.webglLow.gl.canvas = canvasMock;
  figure.canvasHigh = canvasMock;
  figure.htmlCanvas = htmlCanvasMock;
  figure.isTouchDevice = false;
  figure.draw2DLow = new DrawContext2D(definition.width, definition.height);
  figure.draw2DHigh = new DrawContext2D(definition.width, definition.height);
  figure.draw2D = figure.draw2DLow;
  figure.shapesLow = figure.getShapes(false);
  figure.shapesHigh = figure.getShapes(true);
  figure.shapes = figure.shapesLow;
  figure.primitives = figure.shapes;
  // figure.equationLow = figure.getEquations(false);
  // figure.equationHigh = figure.getEquations(true);
  figure.equation = figure.equationLow;
  figure.collectionsLow = figure.getObjects(false);
  figure.collectionsHigh = figure.getObjects(true);
  figure.collections = figure.collectionsLow;
  figure.createFigureElements();
  figure.defaultLineWidth = 0.01;
  figure.primitives.defaultLineWidth = 0.01;
  // needed as the first element needs to be set with the space Transforms
  figure.initElements();
  figure.mock = {
    initialTime: 0,
    duration: 0,
    previousTouchPoint: new Point(0, 0),
    timersBeforeDraw: true,
    timeStep: (deltaTimeInSecondsIn, frameTimeIn = null, draw = true) => {
      const { duration, initialTime } = figure.mock;
      const deltaTimeInSeconds = round(deltaTimeInSecondsIn, 8);
      let frameTime = deltaTimeInSeconds;
      // let deltaTime = 0;
      if (frameTimeIn != null && frameTimeIn < deltaTimeInSeconds) {
        frameTime = frameTimeIn;
        // deltaTime = frameTime;
      }
      let deltaTime = frameTime;

      // let delta = Math.min(deltaTimeInSeconds, frameTime);
      let lastTime = 0;
      deltaTime = round(deltaTime, 8);
      while (deltaTime <= round(deltaTimeInSeconds, 8)) {
        const newNow = round((duration + deltaTime + initialTime) * 1000, 8);
        global.performance.now = () => newNow;
        if (figure.mock.timersBeforeDraw) {
          jest.advanceTimersByTime(round((deltaTime - lastTime) * 1000, 8));
        }
        if (draw) {
          figure.draw(round(newNow / 1000, 8));
        }
        if (!figure.mock.timersBeforeDraw) {
          jest.advanceTimersByTime(round((deltaTime - lastTime) * 1000, 8));
        }
        lastTime = deltaTime;
        if (deltaTime === deltaTimeInSeconds || frameTime === 0) {
          deltaTime += 1;
        } else if (deltaTime + frameTime > deltaTimeInSeconds) {
          deltaTime = deltaTimeInSeconds;
        } else {
          deltaTime += frameTime;
        }
        deltaTime = round(deltaTime, 8);
      }
      figure.mock.duration += deltaTimeInSeconds;
    },
    touchDown: (figurePosition) => {
      const p = getPoint(figurePosition);
      figure.touchDownHandler(p);
      figure.mock.previousTouchPoint = p;
    },
    touchUp: () => {
      figure.touchUpHandler();
    },
    touchMove: (figurePosition) => {
      const p = getPoint(figurePosition);
      figure.touchMoveHandler(figure.mock.previousTouchPoint, p);
      figure.mock.previousTouchPoint = p;
    },
  };
  figure.timeKeeper.reset();
  return figure;
}

