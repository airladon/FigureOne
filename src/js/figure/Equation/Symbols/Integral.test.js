// import {
//   Point,
// } from '../../../../tools/g2';
import {
  round,
} from '../../../tools/math';
import makeFigure from '../../../__mocks__/makeFigure';

describe('Equation Functions - Integral', () => {
  let figure;
  let eqn;
  let functions;
  beforeEach(() => {
    figure = makeFigure();
    functions = {
      symbols: () => {
        figure.addElement({
          name: 'eqn',
          method: 'equation',
          options: {
            color: [0, 0.95, 0, 1],
            elements: {
              a: 'a',
              s1: {
                symbol: 'int',
                color: [0.95, 0, 0, 1], // override default equation color
                lineWidth: 0.01,        // lineWidth
                sides: 20,              // sides in integral s-curve
                width: null,            // symbol width
                tipWidth: null,         // s-curve tip width
                draw: 'static',         // or 'dynamic'
                staticHeight: 'first',  // or number (only use if draw = static)
                serif: true,            // serifs on s-curve
                num: 2,                 // number of s-curves
                type: 'generic',        // or 'line' for line integral symbol
                serifSides: 10,         // sides on serifs
                lineIntegralSides: 20,  // sides on line integral symbol
              },
              s2: {
                symbol: 'int',
                color: [0.95, 0, 0, 1], // override default equation color
                lineWidth: 0.01,        // lineWidth
                sides: 20,              // sides in integral s-curve
                width: null,            // symbol width
                tipWidth: null,         // s-curve tip width
                draw: 'dynamic',        // or 'dynamic'
                serif: false,           // serifs on s-curve
                num: 1,                 // number of s-curves
                type: 'line',           // or 'line' for line integral symbol
                serifSides: 10,         // sides on serifs
                lineIntegralSides: 20,  // sides on line integral symbol
              },
            },
            forms: {
              s1: { content: { int: ['s1', 'a'] } },
              s2: { content: { int: ['s2', 'a'] } },
            },
          },
        });
        figure.initialize();
        eqn = figure.elements._eqn;
      },
    };
  });
  test('Symbol 1', () => {
    functions.symbols();
    eqn.showForm('s1');
    figure.setFirstTransform();
    expect(round(eqn._a.transform.mat)).toMatchSnapshot();
    expect(round(eqn._s1.transform.mat)).toMatchSnapshot();
  });
  test('Symbol 2', () => {
    functions.symbols();
    eqn.showForm('s2');
    figure.setFirstTransform();
    expect(round(eqn._a.transform.mat)).toMatchSnapshot();
    expect(round(eqn._s2.transform.mat)).toMatchSnapshot();
  });
});
