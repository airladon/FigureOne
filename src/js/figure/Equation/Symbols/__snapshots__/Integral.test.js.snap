// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Equation Functions - Integral Symbol 1 1`] = `
Array [
  0.7,
  0,
  0,
  0.14845,
  0,
  0.7,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Integral Symbol 1 2`] = `
Array [
  0.14145,
  0,
  0,
  0,
  0,
  0.2226,
  0,
  -0.0756,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Integral Symbol 2 1`] = `
Array [
  0.7,
  0,
  0,
  0.11105,
  0,
  0.7,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Integral Symbol 2 2`] = `
Array [
  0.10405,
  0,
  0,
  0,
  0,
  0.2226,
  0,
  -0.0756,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;
